app/src/main/java/org/supla/android/widget/lightswitch/LightSwitchCommandWorker.kt

```diff
@@ -0,0 +1,107 @@
+package org.supla.android.widget.lightswitch
+/*
+ Copyright (C) AC SOFTWARE SP. Z O.O.
+
+ This program is free software; you can redistribute it and/or
+ modify it under the terms of the GNU General Public License
+ as published by the Free Software Foundation; either version 2
+ of the License, or (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program; if not, write to the Free Software
+ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ */
+
+import android.appwidget.AppWidgetManager
+import android.content.Context
+import android.os.Handler
+import android.os.Looper
+import android.widget.Toast
+import androidx.work.Worker
+import androidx.work.WorkerParameters
+import org.supla.android.R
+import org.supla.android.SuplaApp
+import org.supla.android.lib.SuplaClient
+import org.supla.android.lib.SuplaConst
+import org.supla.android.widget.SwitchPreferences
+
+const val ARG_TURN_ON = "ARG_TURN_ON"
+
+/**
+ * Worker which is implemented for turning on/off light switches. It supports following channel functions:
+ * light switch [SuplaConst.SUPLA_CHANNELFNC_LIGHTSWITCH],
+ * dimmer [SuplaConst.SUPLA_CHANNELFNC_DIMMER],
+ * RGB lightning [SuplaConst.SUPLA_CHANNELFNC_RGBLIGHTING],
+ * dimmer with RGB lightning [SuplaConst.SUPLA_CHANNELFNC_DIMMERANDRGBLIGHTING]
+ */
+class LightSwitchCommandWorker(appContext: Context, workerParams: WorkerParameters) : Worker(appContext, workerParams) {
+
+    val handler = Handler(Looper.getMainLooper())
+
+    override fun doWork(): Result {
+        val suplaClient = getSuplaClient()
+        if (suplaClient == null) {
+            showToast(R.string.widget_light_switch_command_error)
+            return Result.failure()
+        }
+
+        val widgetIds: IntArray? = inputData.getIntArray(AppWidgetManager.EXTRA_APPWIDGET_IDS)
+        if (widgetIds == null || widgetIds.size != 1) {
+            showToast(R.string.widget_light_switch_command_error)
+            return Result.failure()
+        }
+
+        val configuration = SwitchPreferences(applicationContext).getWidgetConfiguration(widgetIds[0])
+        if (configuration == null) {
+            showToast(R.string.widget_light_switch_command_error)
+            return Result.failure()
+        }
+
+        showToast(R.string.widget_light_switch_command_started)
+
+        if (configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_LIGHTSWITCH) {
```

Comment:
With on/off for RGBW, it is enough to pass the brightness 100 for On and 0 for Off. The dimmer, if it sees the "OnOff" flag, will change 100 to the value memorized before switching it off.   

     if (configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_LIGHTSWITCH) {
            suplaClient.turnOnOff(applicationContext, inputData.getBoolean(ARG_TURN_ON, false),
                    configuration.channelId, false, configuration.channelFunction, false)
        } else if (configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_DIMMER
                || configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_DIMMERANDRGBLIGHTING
                || configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_RGBLIGHTING ) {
            suplaClient.setRGBW(configuration.channelId, configuration.channelColor, 100, 100, true)
        }